K8SPS-624 Documented point-in-time recovery support (#198)
percona/k8sps-docs: docs/backups-pitr.md

# Point in time recovery

!!! admonition ""

    This feature is in the tech preview stage. The behavior can change in future releases.

A base backup captures your data at a single moment in time. Restoring from such a backup is enough in many cases. However, if you need to undo a bad migration, recover right before someone dropped the wrong table, or meet a tighter recovery point, a base backup alone won't give you that level of precision.

> To address these use cases, you must be able to restore the database to a specific moment or to a specific transaction. To do that means to apply the binary logs generated after the backup on top of it, because they contain all subsequent changes. This process is called point-in-time recovery and is available in the Operator starting with v 1.1.0.

Reviewer: ```suggestion
To address these use cases, you must be able to restore the database to a specific moment or to a specific transaction. To do that means to apply the binary logs generated after the backup on top of it, because they contain all subsequent changes. This process is called point-in-time recovery and is available in the Operator starting with v1.1.0.
```

Reviewer: this is not applied

This feature is in the tech preview stage. We don't recommend using it in production yet, but we encourage you to try it out and share your feedback.

## How the Operator performs point‑in‑time recovery

The Operator implements point-in-time recovery in two stages:

* Restores a base physical backup taken with `PerconaXtraBackup`
* Replays binary log on top of the base backup to bring the database to the desired moment or transaction. A binary log records all changes made to the database, such as updates, inserts, and deletes.
  

> To speed up recovery, the Operator converts binary logs into relay logs and applies them concurrently using MySQL's SQL replication thread. This approach reuses MySQL replication capabilities and significantly improves performance.

Reviewer: this is not true anymore—sorry proposal wasn't updated due to time pressure.

we are decoding binlogs using `mysqlbinlog` and apply them using regular `mysql` client.

## Recovery modes

You can restore the database up to a specific moment in time or up to a specific transaction. You specify the desired mode in the `PerconaServerMySQLRestore` object:

| Mode | What you specify | How it works| Typical use case |
| ---- | ----------------- | ----------- |----------------|

> | **GTID** | A GTID set | The Operator restores the database **just before** that transaction. The specified transaction is not included.| Precise, replication-friendly recovery when you know the GTID to stop before|

Reviewer: specified gtid set is included, this statement is not accurate

> | **Date / time** | A timestamp | The Operator restores the database **just before** the last transaction that was written before specified timestamp | Use when wall-clock time is easier than GTIDs |

Reviewer: the operator restores the databases to have all transactions until the target date

## Binary log collection

> The Operator requires binary logs (binlogs) for point-in-time recovery. The Operator collects them using the [Percona Binarylog Server :octicons-link-external-16:](https://github.com/Percona-Lab/percona-binlog-server) command‑line utility. The Binarylog Server connects to MySQL as a replication client and streams binlogs either to a cloud storage or to the filesystem. It can resume collection automatically after being interrupted or stopped.

Reviewer: ```suggestion
The Operator requires binary logs (binlogs) for point-in-time recovery. The Operator collects them using the [Percona Binlog Server :octicons-link-external-16:](https://github.com/Percona-Lab/percona-binlog-server) command‑line utility. The Binlog Server connects to MySQL as a replication client and streams binlogs either to a cloud storage or to the filesystem. It can resume collection automatically after being interrupted or stopped.
```

> The Operator requires binary logs (binlogs) for point-in-time recovery. The Operator collects them using the [Percona Binarylog Server :octicons-link-external-16:](https://github.com/Percona-Lab/percona-binlog-server) command‑line utility. The Binarylog Server connects to MySQL as a replication client and streams binlogs either to a cloud storage or to the filesystem. It can resume collection automatically after being interrupted or stopped.

Reviewer: keep in mind that storing binlogs on filesystem is not supported via operator

Author: Updated

```mermaid
flowchart LR
        PS["Percona Server<br/>primary"]
        BS["Binlog server <br/> replication client <br/> GTID mode"]
        S3["S3-compatible<br/>storage"]
        PS -->|"binary logs"| BS
        BS -->|"archived binlogs"| S3
```

To enable binlog collection, configure the `spec.backup.pitr` section in your Custom Resource:

```
spec:
  backup:
    pitr:
      enabled: true
      binlogServer:
        storage:
          s3:
            bucket: my-binlogs
            credentialsSecret: my-s3-secret
            region: us-east-1
            endpointURL: https://s3.amazonaws.com
```

> After you apply the changes, the Operator starts Percona Binary log Server Pod and enforces GTID mode on the cluster. The Pod continuously collects binlogs and stores them in the configured location.

Reviewer: ```suggestion
After you apply the changes, the Operator starts Percona Binlog Server Pod and enforces GTID mode on the cluster. The Pod continuously collects binlogs and stores them in the configured location.
```

> After you apply the changes, the Operator starts Percona Binary log Server Pod and enforces GTID mode on the cluster. The Pod continuously collects binlogs and stores them in the configured location.

Reviewer: what do you mean enforces GTID mode? GTID mode is already enforced in all deployments via operator

Author: Updated, removed it

## Point-in-time recovery workflow

Let's have a closer look at how point-in-time recovery works:

When you trigger a restore with a point-in-time recovery target, the Operator does the following:

1. Execs into the Binlog server Pod and searches for required binlogs
2. Pauses the cluster
3. Restores the base backup
4. Runs the point-in-time recovery Job that:
   

>     * Starts a temporary `mysql` instance

Reviewer: ```suggestion
    * Starts a temporary `mysqld` instance
```

    * Copies binlogs from the storage and converts them to relay logs
    * Configures MySQL to read and operate only with the downloaded relay logs using the `CHANGE REPLICATION SOURCE TO RELAY_LOG_FILE=..., SOURCE_HOST='dummy'` statement
    * Starts the SQL thread and applies relay logs until the GTID or position. The SQL thread uses the default number of `replica_parallel_workers` for concurrent processing

>     * Stops the replication thread and resets the replica information

Reviewer: - Fetches binlogs from the storage and apply them into `mysql` client on the fly

Reviewer: the other description is not accurate anymore

>     * Shuts down the temporary `mysql` instance

Reviewer: ```suggestion
    * Shuts down the temporary `mysqld` instance
```

  
5. Unpauses the cluster.

For how to configure and start a point-in-time recovery, refer to the [Make a restore with point-in-time recovery on the same cluster](backups-restore.md#restore-with-point-in-time-recovery) and [Make a restore with point-in-time recovery on the new cluster](backups-restore-to-new-cluster.md#restore-with-point-in-time-recovery) tutorials.

With point-in-time recovery, you get finer control over when you come back online, which improves recovery point objective (RPO)

## Implementation specifics

1. Point-in-time recovery is supported for both asynchronous and group replication types.

> 2. The Binarylog server is deployed with the number of Pods restricted to 1. This is because it connects to MySQL as a replication client with a specific server ID and only one instance can connect to the database with the same ID. Any number you set for the `size` in the Custom Resource will be ignored.

Reviewer: ```suggestion
2. The Binlog Server is deployed with the number of Pods restricted to 1. This is because it connects to MySQL as a replication client with a specific server ID and only one instance can connect to the database with the same ID. Any number you set for the `size` in the Custom Resource will be ignored.
```

> 3. Enabling the Binarylog Server enforces GTID mode on the cluster, which is required for replaying relay logs.

Reviewer: GTID mode was already enforced by default

Reviewer: we need to say GTID mode is required but it's already enabled by default

> 4. The cluster is paused during point-in-time recovery. The Operator starts a temporary `mysql` which that operates on the data PVC.

Reviewer: ```suggestion
4. The cluster is paused during point-in-time recovery. The Operator starts a temporary `mysqld` which that operates on the data PVC.
```

> 5. For the restore up to a specific transaction, the Operator restores all transactions except the last one that was written before the specified timestamp. that last transaction is not included.

Reviewer: this is not true anymore-can be removed

> 6. The Binarylog Server must be running before the restore begins. The reconciler locates the required binlogs before pausing the cluster.

Reviewer: ```suggestion
6. The Binlog Server must be running before the restore begins. The reconciler locates the required binlogs before pausing the cluster.
```

> 7. When the cluster is **paused**, point-in-time-related behavior is effectively **disabled** so the Operator does not leave binlog infrastructure running in a way that contradicts the paused state.

Reviewer: this is not true anymore-can be removed

> ## Known limitations

Reviewer: More limitations:

- If `operator` user password is different than the current one in the base backup, recovery will fail. You need to take a new full backup after changing the operator user password.
- Due to updates done by mysql-shell it's possible to see errors like below during the restore:
```
2026-04-15T10:58:08.911878Z 11 [ERROR] [MY-010584] [Repl] Replica SQL: Could not execute Update_rows event on table mysql_innodb_cluster_metadata.instances; Can't find record in 'instances', Error_code: 1032; handler error HA_ERR_KEY_NOT_FOUND; the event's source log FIRST, end_log_pos 162397, Error_code: MY-001032
```

If you want to ignore these errors you need to use `force=true` in restore like:
```
apiVersion: ps.percona.com/v1
kind: PerconaServerMySQLRestore
metadata:
  name: restore1
spec:
  clusterName: cluster1
  backupName: backup1
  pitr:
    force: true
    type: date
    date: "2026-04-16 21:12:00"
```

`force=true` will run mysql client with `--force` flag and this will silently ignore all SQL errors during binlog replay. This might result in data loss. 

- pitr job retries are not idempotent. If recovery fails after restoring the full backup, job won't restore the full backup again to start from a clean state. It's recommended to use `spec.backup.backoffLimit=0` in the cr.yaml

> The Binarylog server supports only AWS S3 and S3-compatible storage services for streaming binlogs to. Use the same operational practices as for backup buckets: credentials via Secrets, correct endpoint URL (including scheme where required), region, and TLS options consistent with your environment.

Reviewer: ```suggestion
The Binlog Server supports only AWS S3 and S3-compatible storage services for streaming binlogs to. Use the same operational practices as for backup buckets: credentials via Secrets, correct endpoint URL (including scheme where required), region, and TLS options consistent with your environment.
```

1. If something fails mid-restore, use the same discipline as with any restore: inspect **`PerconaServerMySQLRestore` status**, the **restore and PITR Jobs**, and [Troubleshoot backups and restores](debug-backup-restore.md).

> 2. You can only perform point-in-time recovery from a backup specified using the `backupName` field. Point-in-time recovery is currently not supported when restoring from a backup specified via `dataSource`. This improvement is planned for future releases.

Reviewer: this is not true. you can use `backupSource` (not `dataSource`) with base backup.

what you can't do is setting backupSource for binary log storage.

* If the Operator user password is different from the password saved in the base backup, point-in-time recovery will fail. You must take a new full backup after changing the Operator user password to ensure point-in-time recovery works.

* During point-in-time recovery, updates made by `mysql-shell` can result in replication errors such as:
  
   ```
   2026-04-15T10:58:08.911878Z 11 [ERROR] [MY-010584] [Repl] Replica SQL: Could not execute Update_rows event on table mysql_innodb_cluster_metadata.instances; Can't find record in 'instances', Error_code: 1032; handler error HA_ERR_KEY_NOT_FOUND; the event's source log FIRST, end_log_pos 162397, Error_code: MY-001032
   ```
  
   If you want to ignore such errors, add `force: true` to the PITR section of your restore spec:
  
   ```yaml
   apiVersion: ps.percona.com/v1
   kind: PerconaServerMySQLRestore
   metadata:
     name: restore1
   spec:
     clusterName: cluster1
     backupName: backup1
     pitr:
       force: true
       type: date
       date: "2026-04-16 21:12:00"
   ```

>   The `force: true` option uses the `--force` flag with the MySQL client and will silently ignore all SQL errors during binlog replay. **Warning:** This might result in data loss if underlying replication or data integrity errors are ignored.

Reviewer: could you please document force option in restore-cr doc?
